core/appeal/service.go

```diff
@@ -486,6 +493,367 @@ func validateAppealOnBehalf(a *domain.Appeal, policy *domain.Policy) error {
+	}
+
+	if err := s.addCreatorDetails(ctx, appeal, policy); err != nil {
+		return fmt.Errorf("getting creator details: %w", err)
+	}
+
+	// mark previous approvals as stale
+	for _, approval := range existingAppeal.Approvals {
+		approval.IsStale = true
+		err := s.approvalService.UpdateApproval(ctx, approval)
+		if err != nil {
+			return fmt.Errorf("couldn't update old approval(%s): %w", approval.ID, err)
+		}
+	}
+
+	// create new approval
+	appeal.Revision = existingAppeal.Revision + 1
+	if err := appeal.ApplyPolicy(policy); err != nil {
+		return err
+	}
+
+	if err := appeal.AdvanceApproval(policy); err != nil {
+		return fmt.Errorf("initializing approvals: %w", err)
+	}
+	appeal.Policy = nil
+
+	notifications := []domain.Notification{}
+	for _, approval := range appeal.Approvals {
+		if approval.Index == len(appeal.Approvals)-1 && (approval.Status == domain.ApprovalStatusApproved || appeal.Status == domain.AppealStatusApproved) {
+			newGrant, revokedGrant, err := s.prepareGrant(ctx, appeal)
+			if err != nil {
+				return fmt.Errorf("preparing grant: %w", err)
+			}
+			newGrant.Resource = appeal.Resource
+			appeal.Grant = newGrant
+			if revokedGrant != nil {
+				if _, err := s.grantService.Revoke(ctx, revokedGrant.ID, domain.SystemActorName, revokedGrant.RevokeReason,
+					grant.SkipNotifications(),
+					grant.SkipRevokeAccessInProvider(),
+				); err != nil {
+					return fmt.Errorf("revoking previous grant: %w", err)
+				}
+			} else {
+				if err := s.GrantAccessToProvider(ctx, appeal, opts...); err != nil {
+					return fmt.Errorf("granting access: %w", err)
+				}
+			}
+
+			notifications = append(notifications, domain.Notification{
+				User: appeal.CreatedBy,
+				Labels: map[string]string{
+					"appeal_id": appeal.ID,
+				},
+				Message: domain.NotificationMessage{
+					Type: domain.NotificationTypeAppealApproved,
+					Variables: map[string]interface{}{
+						"resource_name": fmt.Sprintf("%s (%s: %s)", appeal.Resource.Name, appeal.Resource.ProviderType, appeal.Resource.URN),
+						"role":          appeal.Role,
+						"account_id":    appeal.AccountID,
+						"appeal_id":     appeal.ID,
+						"requestor":     appeal.CreatedBy,
+					},
+				},
+			})
+
+			notifications = addOnBehalfApprovedNotification(appeal, notifications)
+		}
+	}
+
+	if err := s.repo.UpdateByID(ctx, appeal); err != nil {
+		return fmt.Errorf("error saving appeal to db: %w", err)
+	}
+
+	log, err := getAuditLog(existingAppeal, appeal)
+	if err != nil {
+		return err
+	}
+
+	if err := s.auditLogger.Log(ctx, AuditKeyUpdateAppeal, log); err != nil {
+		s.logger.Error(ctx, "failed to record audit log", "error", err)
+	}
+
+	if len(notifications) > 0 {
+		if errs := s.notifier.Notify(ctx, notifications); errs != nil {
+			for _, err1 := range errs {
+				s.logger.Error(ctx, "failed to send notifications", "error", err1.Error())
+			}
+		}
+	}
+
+	return nil
+}
+
+func getAuditLog(oldAppeal, newAppeal *domain.Appeal) (string, error) {
```

Comment:
can we transform both `oldAppeal` and `newAppeal` to json and then use `diff.GetChangelog` to get the diff?

Reply:
yes, but that will create diff for other fields as well that are not required in `auditLog` like grants, approvals, updatedAt